src/app/nodetool/qgsmaptoolnodetool.cpp

```diff
@@ -17,463 +17,164 @@
-        {
-          mRubberBands[i]->setTranslationOffset( offset.x(), offset.y() );
-        }
-        return;
-      }
-
-      // move points
-      QList<QgsVertexEntry*> &vertexMap = mSelectedFeature->vertexMap();
-      for ( int i = 0; i < vertexMap.size(); i++ )
-      {
-        if ( !vertexMap[i]->isSelected() )
-          continue;
-
-        QgsPoint p = toMapCoordinates( vlayer, vertexMap[i]->point() ) + offset;
+      int atVertex, beforeVertex, afterVertex;
+      double dist;
 
-        mRubberBands[vertexMap[i]->rubberBandNr()]->movePoint( vertexMap[i]->rubberBandIndex(), p );
-
-        if ( vertexMap[i]->rubberBandIndex() == 0 )
-        {
-          mRubberBands[vertexMap[i]->rubberBandNr()]->movePoint( 0, p );
-        }
-      }
-
-      // topological editing
-      offset = posMapCoord - mPosMapCoordBackup;
-      for ( int i = 0; i < mTopologyRubberBand.size(); i++ )
-      {
-        for ( int pointIndex = 0; pointIndex < mTopologyRubberBand[i]->numberOfVertices(); pointIndex++ )
-        {
-          if ( mTopologyRubberBandVertexes[i]->contains( pointIndex ) )
-          {
-            const QgsPoint* point = mTopologyRubberBand[i]->getPoint( 0, pointIndex );
-            if ( point == 0 )
-            {
-              break;
-            }
-            mTopologyRubberBand[i]->movePoint( pointIndex, *point + offset );
-          }
-        }
-      }
-
-      mPosMapCoordBackup = posMapCoord;
-    }
-  }
-  else
-  {
-    if ( !mSelectionRectangle )
-    {
-      mSelectionRectangle = true;
-      mSelectionRubberBand = new QRubberBand( QRubberBand::Rectangle, mCanvas );
-      mRect = new QRect();
-      mRect->setTopLeft( mPressCoordinates );
+      QgsPoint closestLayerVertex = mSelectedFeature->geometry()->closestVertex( e->mapPoint(), atVertex, beforeVertex, afterVertex, dist );
+      mSelectedFeature->selectVertex( atVertex );
     }
-    mRect->setBottomRight( e->pos() );
-    QRect normalizedRect = mRect->normalized();
-    mSelectionRubberBand->setGeometry( normalizedRect );
-    mSelectionRubberBand->show();
   }
 }
 
-void QgsMapToolNodeTool::canvasPressEvent( QMouseEvent * e )
+void QgsMapToolNodeTool::canvasMapMoveEvent( QgsMapMouseEvent* e )
 {
-  QgsDebugCall;
+  if ( !mSelectedFeature || !mSelectedFeature->hasSelection() )
+  {
+    return;
+  }
 
-  mClicked = true;
-  mPressCoordinates = e->pos();
-  if ( !mSelectedFeature )
+  QgsVectorLayer* vlayer = mSelectedFeature->vlayer();
+  if ( !vlayer )
   {
-    QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( mCanvas->currentLayer() );
-    if ( !vlayer )
-      return;
+    return;
+  }
 
-    mSelectAnother = false;
-    QgsPointLocator::Match m = mCanvas->snappingUtils()->snapToCurrentLayer( e->pos(), QgsPointLocator::Vertex | QgsPointLocator::Edge );
-    if ( !m.isValid() )
+  if ( mMoveRubberBands.empty() )
+  {
+    QgsGeometryRubberBand* rb = new QgsGeometryRubberBand( mCanvas, mSelectedFeature->geometry()->type() );
```

Comment:
QgsSelectedFeature (in contrast to QgsFeature) does not have a constGeometry(). But yes, for 424,431 and the two other occurences below, it is better to use constGeometry().

>to avoid the detach

afaik detach is only called when a non-const operation is really called on that object.